File: src/ZimCreator.ts
import {statSync} from 'fs';
import * as rimraf from 'rimraf';

import {ZimArticle, ZimCreatorWrapper} from './zim';

export interface ZimCreatorOpts {
  fileName: string;
  welcome?: string;
  fullTextIndexLanguage?: string;
  minChunkSize?: number;
}

export interface ZimMetadata {
  Counter?: string, Creator?: string, Date?: string, Description?: string,
      Flavour?: string, Language?: string, Name?: string, Publisher?: string,
      Tags?: string, Title?: string,
}


class ZimCreator {
  tmpDir: string;
  _creator: any;
  fileName: string;
  articleCounter: {[mimeType: string]: number} = {};

  constructor(opts: ZimCreatorOpts, metadata: ZimMetadata = {}) {
    // {fileName, welcome = '', fullTextIndexLanguage = '', minChunkSize =
    // 2048}, metadata = {}) {
    this.fileName = opts.fileName;
    this.articleCounter = {};
    this.tmpDir = this.fileName.split('.').slice(0, -1).join('.') + '.tmp';
    const metadataDefaults = {
      Date: (new Date()).toJSON().split('T')[0],
      Language: 'eng',
    };
    metadata = Object.assign(metadataDefaults, metadata);

    try {
      const statRes = statSync(this.tmpDir);
      console.info(`Tmp directory already exists, deleting [${this.tmpDir}]`);
      rimraf.sync(this.tmpDir);
    } catch (err) {
      console.error(err);
    }

    const {
      welcome = '',
      fullTextIndexLanguage = '',
      minChunkSize = 2048,
    } = opts;
    this._creator = new ZimCreatorWrapper({
      fileName: this.fileName,
      mainPage: welcome,
      fullTextIndexLanguage,
      minChunkSize: minChunkSize,
    });

    this.setMetadata(metadata);
  }

  get isAlive() {
    return !!this._creator;
  }

  async addArticle(article: any) {
    if (!this.isAlive) throw new Error(`This Creator has been destroyed`);
    this._creator.addArticle(article);
    if (!article.redirectUrl) {
      this.articleCounter[article.mimeType] =
          this.articleCounter[article.mimeType] || 0;
      this.articleCounter[article.mimeType] += 1
    }
    return article;
  }

  async setMetadata(metadata: ZimMetadata) {
    if (!this.isAlive) throw new Error(`This Creator has been destroyed`);
    const keys = Object.keys(metadata).filter(key => (metadata as any)[key]);
    for (const key of keys) {
      const content = (metadata as any)[key];
      const article = new ZimArticle({url: key, data: content, ns: 'M'});
      await this.addArticle(article);
    }
  }

  async finalise() {
    if (!this.isAlive) throw new Error(`This Creator has been destroyed`);
    const counterString = Object.keys(this.articleCounter)
                              .map(mimeType => `${mimeType} =
                                $ { this.articleCounter[mimeType] }`)
                              .join(';');
    await this.setMetadata({Counter: counterString});
    this._creator.finalise();
    this._creator = null;
  }
};

export {ZimCreator};

File: src/ZimCreator.cc
#include <string>
#include <iostream>
#include <zim/writer/creator.h>

#include "Article.cc"
#include "nbind/api.h"

class OverriddenZimCreator : public zim::writer::Creator
{
public:
	OverriddenZimCreator(std::string mainPage)
		: zim::writer::Creator(true),
		  mainPage(mainPage) {}

	virtual zim::writer::Url getMainUrl()
	{
		return zim::writer::Url('A', mainPage);
	}

	std::string mainPage;
};

class ZimCreatorWrapper
{
public:
	ZimCreatorWrapper(OverriddenZimCreator *creator) : _creator(creator)
	{
	}

	~ZimCreatorWrapper()
	{
		delete _creator;
	}

	static ZimCreatorWrapper *create(std::string fileName, std::string mainPage, std::string fullTextIndexLanguage, int minChunkSize)
	{
		bool shouldIndex = !fullTextIndexLanguage.empty();

		OverriddenZimCreator *c = new OverriddenZimCreator(mainPage); // TODO: consider when to delete this
		c->setIndexing(shouldIndex, fullTextIndexLanguage);
		c->setMinChunkSize(minChunkSize);
		c->startZimCreation(fileName);
		return (new ZimCreatorWrapper(c));
	}

	void addArticle(Article article, nbind::cbFunction &callback)
	{
		try
		{
			ZimArticle a = article.toZimArticle();
			_creator->addArticle(a);
			callback();
		}
		catch (...)
		{
			callback("Failed to add article");
			throw;
		}
	}

	void finalise(nbind::cbFunction &callback)
	{
		_creator->finishZimCreation();
		callback();
		delete this;
	}

	OverriddenZimCreator *_creator;
};

#include "nbind/nbind.h"

#ifdef NBIND_CLASS

NBIND_CLASS(ZimCreatorWrapper)
{
	construct<OverriddenZimCreator *>();
	method(create);
	method(addArticle);
	method(finalise);
}

#endif
